# Module 10: Descriptive Statistics and Outliers
***

## What are statistics?

Statistics deal with collecting informative data, interpreting these data, and drawing conclusions about the data. Statistics are used across multiple disciplines and fuel business and policy decisions.

## Two types of statistics
***
#### Descriptive Statistics
Descriptive statistics use numeric and graphical techniques to summarize the characteristics of a set of data and to identify patterns within the dataset. Descriptive statistics also include graphically summarizing data and presenting the information in a meaningful (easy to interpret) way. 

#### Inferential Statistics
Inferential statistics use samples of data to make generalizations about a larger set of data (or a larger population). Inferential statistics use data to make estimates, decisions, and predictions about patterns and trends identified within a set of data. Inferential statistics are used to make conclusions beyond your immediate set of data. 

## Import Libraries and Data

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats ## new library alert! ##

## SciPy is used for scientific and mathematic computing; it provides functions for stats, 
# data manipulation, and data visualization

df = pd.read_excel("axisdata.xlsx")

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Qualitative vs Quantitative Data
***

#### Quantitative Data
Data that are recorded on a numeric scale (i.e. age, weight, salary, number of empty seats on a plane, credit score etc). Quantitative variables are also called <b>numeric</b> variables. 

#### Qualitative Data
Data that cannot be measured on a numeric scale (i.e. eye color, type of car, blood type, race etc). These data can only be classified into one group of categories. Qualitative variables are also called <b>categorical</b> variables. Qualitative data also includes text and string data. 

In [65]:
## Which variables are qualitative and which variables are quantitative?

df.head()

Fname      Lname Position Gender  Hours Worked SalesTraining  \
0  Jackie    Jackson  Trainee      F            32             Y   
1    Mary  Patterson  Trainee      F            43             N   
2   Tanya      Adams  Trainee      F            28             Y   
3   Tanya  Henderson  Trainee      F            24             Y   
4  Walter   Franklin  Trainee      M            25             Y   

   Years Experience  Cars Sold  
0                 1          3  
1                 1          6  
2                 1          2  
3                 1          5  
4                 1          4

## Describing Qualitative Data
***

Qualitative data is best summarized by frequency tables. 

Frequency tables summarize categorical data into a table showing each category, number of observations that fall within each category, and the relative frequency of each category. 

Relative frequency is the number of specific observations out of the total number of observations, or the percentage of the total sample that fall within a specific group. Both representations of categorical data give insight into how common specific categories are within a set of data. 

<b>Bar charts</b> are ideal for graphically representing qualitative data as they visually show the frequency of observations across various groups. 

#### GRAPHIC SHOULD READ N = 150 !!!

<img src="FreqTable.png">

In [66]:
## Frequencies 

df["Gender"].value_counts()

M    510
F    489
Name: Gender, dtype: int64

In [67]:
## Relative Frequencies 

df["Gender"].value_counts(normalize=True)

M    0.510511
F    0.489489
Name: Gender, dtype: float64

In [68]:
## Multi-variable Frequencies 

pd.crosstab(df["Gender"], df["SalesTraining"], margins=False, normalize=False)

SalesTraining    N    Y
Gender                 
F              212  277
M              201  309

# { Exercise 1 }

    1. Import the "pokemon.csv" file; name the dataset 'poke'. Preview the first 5 rows. 
    2. Determine the frequencies of unique pokemon types (variable name: Type 1) in this dataset
    3. Determine the relative frequencies of the unique pokemon types (variable name: Type 1) in this dataset
    4. Create a crosstabs table that shows the frequency of pokemon stage (variable name: Stage) and legendary status (variable name: Legendary). 

## Describing Quantitative Data

There are many more techniques for describing quantitative data compared to qualitative. When summarizing quantitative data, the objective is to get a sense of how the data is clustered and the variation in numeric values. 

***

### Measures of Central Tendency

Measures of Central Tendency are estimates of what a "typical" value/data point is within your set of data. A measure of central tendency is a number around which an entire sample of data is spread. We can describe this central position using several statistics, including the mean (average), the median, and the mode. 

Measures of central tendency can be visualized with a frequency distribution (histogram). A frequency distribution is a description of values and how often each value occurs within your dataset. When you graph a frequency distribution a pattern/shape should give insight into which values occured the most/least. 

<img src="NormalDist.jpg">

***

#### Mean

The mean, which is the most commonly known and used measure of central tendency, is the arithmetic average of a set of numeric values. The mean is calculated by adding all the values together and dividing by the total number of values in the set of values. In specific sets of data, the mean is the central point in the frequency distribution aka the hump. 

In [69]:
## Calculate the Mean of a column 

df["Hours Worked"].mean()

33.727727727727725

#### Median 

The median is the value which divides the data into two equal parts when the data is arranged from least to greatest. The median will be the middle value if the number of values is odd, otherwise, it will be the average of the two middle values if the number of values is even. 

In [70]:
## Calculate the Median of a column 

df["Hours Worked"].median()

34.0

#### Mode 

The mode is the value that appears most frequently in a sample of data, or the value that has the highest frequency. It is possible for a set of data to have no mode, and it is possible for a set of data to have multiple modes. 

In [71]:
## Calculate the Mode of a column 

df["Hours Worked"].mode()

0    38
dtype: int64

# { Exercise 2 }

    1. Using the pokemon dataset ('poke'), calculate the mean attack value (Attack) in the dataset.  
    2. Determine the median defense value (Defense). Write a single sentence on what your findings mean. 
    3. Determine the mode of the speed (Speed) column. Write a single sentence on what your findings mean. 

### Data Distribution and Measurement Sensitivity
***

### Normally distributed data vs. non-normally distributed data

Several statistics are suited for data that follows a <b>normal distribution</b>. When visualized, normally distributed data lies in a symmetrical pattern (bell-shape) where the majority of the data surrounds the mean with decreasing amounts evenly distributed to the left and the right. 

While this is the ideal, real-world data is not always normally distributed. Non-normal distribution is possible and there are several statistics that can still be used with non-normally distributed data. It is important to determine how your data is distributed before you embark on certain statistics. 

#### Distribution and Measure of Central Tendency

The shape of the data helps to determine which measure of central tendency should be reported. Some statistics are very sensitive to non-normal distributions, while others are more resilient. Three of the most common shapes to learn are symmetric, left-skewed, and right-skewed. <b>Skew</b> is simply a measure of the asymmetry of the distribution. <b>In the below images, while not shown, the values are increasing from left to right.</b> 

#### Symmetric Data
* mean, median, and mode are all the same here
* no skewness is apparent
* the distribution is described as symmetric

<img src="Symmetrical.jpg">

***

#### Left-skewed Data
* mean < median
* long tail on the left

<img src="LeftSkew.jpg">

***

#### Right-skewed Data
* mean > median
* long tail on the right

<img src="RightSkew.jpg">

***

### Considerations for Shape

When your data is skewed in either direction, you should consider the best measure of central tendency to report. The median is best used when you have skewed data. This is because the median is less affected by extreme values. 

If your data is skewed, this could be an indication that your data has extreme high or low values that are contributing to the skewness. Later in this lesson, we will talk about how to handle extreme values. 

### Measures of Variability (or Spread)

Measures of Variability, or spread, describe the amount of dispersion in your data sample. In other words, measures of variability, describe how spread out values are. For example, in a sample of 100 students, the average test score is 75. However, not all students will have gotten a 75 - their scores will be spread out above and below the average, some will be higher and some lower. To describe the spread of our data, we can use several statistics, including range, standard deviation, percentiles, and quartiles.

***

#### Range

The range is the difference between the lowest and highest value in a set of data. For example, if the maximum value is 10 and the minimum value is 2, the range is 8. The larger the range, the larger the difference between the highest and lowest values. 

In [2]:
## Calculate the range of values in a column

hrs_range = df['Hours Worked'].max() - df['Hours Worked'].min()

print(hrs_range)

28


#### Standard Deviation (SD)

The SD is a measurement of the average distance between each data value and the mean. You can consider SD as a measure of how spread out the data is from the mean. A low SD indicates that the data points are clustered around the mean, while a high SD indicates the data points are spread out over a wider range of values. 

<img src="https://d20khd7ddkh5ls.cloudfront.net/high_low_standard_deviation.png">

In [3]:
## Calculate the Standard Deviation of values in a column

print("Mean hours worked by all employees:", df["Hours Worked"].mean())

df["Hours Worked"].std()

Mean hours worked by all employees: 33.727727727727725


8.223453795492198

#### Percentiles

Percentiles give insight into the relation of one data point to other data points within the same dataset. When the data values are arranged from lowest to highest, percentiles represent the position of a value within the list of values. For example, if a value is at the 50th percentile - 50% of the data falls below the value, and 50% fall above the value. If a value is at the 85th percentile - 85% of the data falls below the value, and 15% fall above the value. 

In [4]:
## Calculate percentiles of values in a column
## To determine the value that falls at a specific percentile
## np.percentile(data/column, percentile value)

np.percentile(df['Hours Worked'], 85)

## Determine which value falls at the 85th percentile? 
## 85% of the values in Hours Worked are below this value, 15% are above this value. 

44.0

#### Quartiles 

Quartiles are values that divide a set of data into equal quarters (when the data is ordered from lowest to highest). There are <b>three</b> quartile values that divide the data into four parts (see below). 

<img src="quar.png">

* The first quartile (Q1) is at the 25th percentile (25% below, 75% above)
* The second quartile (Q2) is at the 50th percentile (50% below, 50% above, median)
* The third quartile (Q3) is at the 75th percentile (75% below, 25% above)

<b>Quartiles can be used to calculate the Interquartile Range (IQR) - which can be used to handle outliers in your dataset.</b>

In [75]:
## Determine the three quartiles of a set of data
## A quantile is a series of values, or cut points, that divide a set of data into equal parts 
## A 'quartile' is a specical kind of quantile that divides the data into four equal parts

df['Hours Worked'].quantile(.25) 

#Q1 = .25
#Q2 = .50
#Q3 = .75

27.0

#### All of the Above 

The describe function is perfect for outputting multiple descriptive statistics at once. This will save you multiple steps if you want to look at one or several quantitative variables. 

In [76]:
## Use the describe function to view multiple statistics at the same time

df.describe()

Hours Worked  Years Experience   Cars Sold
count    999.000000        999.000000  999.000000
mean      33.727728          3.026026    3.922923
std        8.223454          1.394709    1.527000
min       20.000000          1.000000    1.000000
25%       27.000000          2.000000    3.000000
50%       34.000000          3.000000    4.000000
75%       41.000000          4.000000    5.000000
max       48.000000          5.000000    7.000000

In [77]:
## Use the describe function to isolate information for a single column

df['Hours Worked'].describe()

count    999.000000
mean      33.727728
std        8.223454
min       20.000000
25%       27.000000
50%       34.000000
75%       41.000000
max       48.000000
Name: Hours Worked, dtype: float64

# { Exercise 3 }

    1. Using the pokemon dataset ('poke'), calculate the range for the HP column. 
    2. What is the mean and the standard deviation for the special attack (SpAtk) column?
    3. For the Attack column, determine which value falls at the 80th percentile? Which value falls at the 15th percentile? For each of these values, write a single sentence describing what this means. 
    4. Calculate the second quartile (or the median) for the special defense (SpDef) column. Then, calculate the median of the same column using the method shown above. Are the values the same? 

## Handling and Removing Outliers

***

Outliers are data points that differ significantly from other data points. Outliers are an important feature of a dataset that should be identified and addressed because extreme values can influence the results of your statistics. 

   * In a sample of 5 students, their ages are 25, 22, 97, 23, 21 - 97 is an outlier because the value differs significantly from the remaining data.
   * Without removing outlier, the mean is: 37.6 and the median is: 23
   * When removing the outlier, the mean is: 22.75 and the median is: 22.5

Determining which values are extreme can be challenging when you have a larger dataset. There are several methods that can be used to strategically determine which values are outliers, we will cover two separate methods below. 

In [5]:
df = pd.read_csv("SalaryData.csv")
df.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  
0          Y      2                    98.3  
1          Y      3                   105.7  
2          N      1                   145.0  
3          Y      3                    75.1  
4          N      2                   131.1

## Method 1 : Using z-scores to detect and remove outliers

In statistics, the <font color=salmon><b>Empirical Rule</b></font> estimates how data, following a normal distribution, will spread around the mean. According to this rule, basically all data values will fall within 3 SD's above or below the mean. The Empirical Rule is also known as the <font color=salmon><b>68-95-99.7 Rule</b></font> because according to this rule, your data will typically fall into the following pattern:

* 68% of the data falls within one SD of the mean

* 95% of the data falls within two SD's of the mean

* 99.7% of the data falls within three SD's of the mean

<img src="SD Bell Curve.jpeg">

#### Empirical Rule Example:

The average weight of an American man is 175 lbs with a SD of 10 lbs. According to the Empirical Rule, we can assume: 

* 68% of American men are between 165 and 185 lbs (or +/- 10 lbs from the mean, or 1 SD)
* 95% of American men are between 155 and 195 lbs (or +/- 20 lbs from the mean, or 2 SD's)
* 99.7% of American men are between 145 and 205 lbs (or +/- 30 lbs from the mean, or 3 SD's)
* Typically, data values that fall outside of this range are considered outliers
_______________________ 
A <font color=salmon><b>z-score</b></font>, also known as a standard score, measures the distance of a value from the mean. In other words, the z-score tells you how many SD's a given value is from the mean. Values that fall above the mean will have a positive (+) z-score, while values that fall below the mean will have a negative (-) z-score. Values that are far from the mean will be considered outliers. You can define a value as an outlier if the z-score greater than +/- 3. 
_______________________ 

#### Steps for Detecting and Removing Outliers with z-scores
* Import the scipy library to make use of the z-score function
* Calculate the z-score for each value within a specific column in your dataset
* Transform the z-score to the absolute value of the z-score (number without the +/- sign; this will make our code simpler)
* Add the z-scores to your dataset as a new column
* Remove all rows that have a z-score greater than 3

In [6]:
## Create a copy of your dataset to filter outliers
dfz = df.copy()

## Check original shape of the dataset
print(dfz.shape)

(30, 10)


In [80]:
## Create a new column to contain the z-scores for the employee salary
## Set the new column equal to the absolute value of the z-scores
## To calculate z-score -- stats.zscore(data/column name)
## To determine absolute value of z-score -- np.abs(data/function/options)

dfz["zscore_salary"] = np.abs(stats.zscore(dfz["Salary"]))

## Preview new column; optional
dfz.head()

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
0  Richard              5.1    66029   41  Female             0  Central   
1     Nick              5.9  8136300   39  Female             0  Central   
2   Morgan              9.5   116969   50  Female             1  Central   
3    Susan              3.2    64445   27    Male             0  Central   
4  Matthew              8.2   113812   47  Female             1  Central   

  CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
0          Y      2                    98.3       0.269844  
1          Y      3                   105.7       3.343557  
2          N      1                   145.0       0.247036  
3          Y      3                    75.1       0.270553  
4          N      2                   131.1       0.248449

In [81]:
## Determine the index locations for the rows with zscores that are greater than "3"
z_outliers = dfz.loc[dfz["zscore_salary"] > 3].index

## Preview list of index values
print(z_outliers)

Int64Index([1, 28], dtype='int64')


In [82]:
## What information can we find at these index locations?

dfz.iloc[[1, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  zscore_salary  
1           Y      3                   105.7       3.343557  
28          Y      3                   116.8       4.100689

In [83]:
## Drop rows with above index values
dfz = dfz.drop(z_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfz.shape)

(28, 11)


## Method 2 : Using the Interquartile Range (IQR) to detect and remove outliers

The <font color=salmon><b>Interquartile Range (IQR)</b></font> is the difference between the first and third quartile in a set of data. The IQR can be considered where the bulk of the data values rest. IQR is calculated by the following equation: <B>IQR = Q3 - Q1</B>
<img src="IQR.png">
_____________________

The IQR can be used to create "fences" or cut-off values that determine which values fall outside of the acceptable data range. For example, we can calculate a lower-limit -- all values falling below this limit will be considered an outlier. The <b>1.5 Rule</b> is a commonly used method for determining the upper and lower cut-off limits. According to this rule, a value is considered an outlier if it falls:

* ... below Q1 - (1.5 x IQR)

* ... above Q3 + (1.5 x IQR)
_____________________

#### 1.5 Rule Example:

A professor is tracking the number of absences for each of their students. They have compiled a list of the number of absences: 

    1, 3, [4], 6, 7, [7], 8, 8, [9], 22, 37 

In this list of data, we can find the following values:

* Q1 = 4
* Q2 (median) = 7
* Q3 = 9

With these values, you can easily calculate the IQR:

* <B>IQR</B> = Q3(9) - Q1(4)
* <B>IQR</B> = 5

Once you have the IQR, you can calculate the upper and lower limits for detecting outliers:

* Upper Limit = Q3(9) + [1.5 * IQR(5)] = <b> 16.5 </b>
* Lower Limit = Q1(4) - [1.5 * IQR(5)] = <b> -3.5 </b>

Values that fall above the upper limit are outliers; values that fall below the lower limit are outliers
_______________________ 

#### Steps for Detecting and Removing Outliers with IQR

* Calculate quartiles
* Calculate IQR
* Determine upper and lower limits
* Drop values less than the lower and/or greater than the upper

In [84]:
## Create a copy of original dataset
dfq = df.copy()

## Check original shape of the dataset
dfq.shape

(30, 10)

In [85]:
## Calculate quartiles
q1 = dfq["Salary"].quantile(.25)
q3 = dfq["Salary"].quantile(.75)

print("Q1:", q1)
print("Q3:", q3)

Q1: 48265.0
Q3: 104512.0


In [86]:
## Calculate the IQR
iqr = q3 - q1

print("IQR:", iqr)

IQR: 56247.0


In [87]:
## Determine outlier fences 
top = q3 + (iqr * 1.5)
bottom = q1 - (iqr * 1.5)


print("Upper Limit:", top)
print("Lower Limit:", bottom)

Upper Limit: 188882.5
Lower Limit: -36105.5


In [88]:
## Determine the index locations for rows that fall outside of outlier fences

iqr_outliers = dfq.loc[(dfq['Salary'] > top) | (dfq['Salary'] < bottom)].index

print("INDEX VALUES:", iqr_outliers)

INDEX VALUES: Int64Index([1, 12, 28], dtype='int64')


In [89]:
## what values can we find at these index locations?

dfq.iloc[[1, 12, 28]]

Employee  YearsExperience   Salary  Age  Gender  RemoteStatus   Region  \
1      Nick              5.9  8136300   39  Female             0  Central   
12   Thomas              1.5   377310   46    Male             1  Central   
28     Bill              7.1  9827300   36  Female             0     East   

   CompanyCar  Shift  PerformanceReviewScore  
1           Y      3                   105.7  
12          Y      1                    63.8  
28          Y      3                   116.8

In [90]:
## Drop rows with above index values
dfq = dfq.drop(iqr_outliers)

## Re-check the shape of the dataframe, how many rows were dropped?
print(dfq.shape)

(27, 10)


# { Module 10 Homework }

1. Import the "babies.xlsx" dataset. See below for information on the columns:

    * bwt - birth weight of newborn baby
    * gestation	- gestation length (weeks)
    * parity - previously pregnant (0 = no; 1 = yes)
    * age - age of mother
    * height - height of mother (inches)	
    * weight - weight of mother (pounds)
    * smoke - smoking status of mother (0 = nonsmoker; 1 = smoker)

In [1]:
import pandas as pd
import numpy as np
import scipy.stats as stats 

df = pd.read_excel("babies.xlsx")

2. Preview the first few rows of the dataset. Complete the following checks for your dataset:

    * output the summary information for the dataset, what types of variables make up this dataset?
    * is there any missing data in this dataset?

In [2]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  123        NaN       0  36.0    69.0   190.0    0.0
4  108      282.0       0  23.0    67.0   125.0    1.0

In [3]:
df.info()

## integers and floats
## There is missing information.

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1236 entries, 0 to 1235
Data columns (total 7 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   bwt        1236 non-null   int64  
 1   gestation  1223 non-null   float64
 2   parity     1236 non-null   int64  
 3   age        1234 non-null   float64
 4   height     1214 non-null   float64
 5   weight     1200 non-null   float64
 6   smoke      1226 non-null   float64
dtypes: float64(5), int64(2)
memory usage: 67.7 KB


3. Handle the missing data in the dataset -- there isn't much, so we can drop all the rows that have at least one missing value. How many rows were dropped from the dataset?

In [4]:
df.dropna(inplace = True)

In [7]:
df.shape

(1174, 7)

In [8]:
1236 - 1174

## 62 rows were removed

62

In [9]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
4  108      282.0       0  23.0    67.0   125.0    1.0
5  136      286.0       0  25.0    62.0    93.0    0.0

In [10]:
df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
3  108      282.0       0  23.0    67.0   125.0    1.0
4  136      286.0       0  25.0    62.0    93.0    0.0

4. There are two qualitative variables in this dataset. What are they? How do you know they are qualitative variables?

In [ ]:
## smoke and parity
## it is data that cannot be measured on a numeric scale they are categorical

5. Replace the values in the qualitative variables with meaningful labels that describe the different groups. Use any method you've learned in previous modules to complete this task. 

In [5]:
df.head()

bwt  gestation  parity   age  height  weight  smoke
0  120      284.0       0  27.0    62.0   100.0    0.0
1  113      282.0       0  33.0    64.0   135.0    0.0
2  128      279.0       0  28.0    64.0    77.0    1.0
4  108      282.0       0  23.0    67.0   125.0    1.0
5  136      286.0       0  25.0    62.0    93.0    0.0

In [11]:
df["parity"].replace([0,1], ["1st Pregnancy", "Multiple Pregnancies"], inplace = True)

In [12]:
df.head()

bwt  gestation         parity   age  height  weight  smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0    0.0
1  113      282.0  1st Pregnancy  33.0    64.0   135.0    0.0
2  128      279.0  1st Pregnancy  28.0    64.0    77.0    1.0
3  108      282.0  1st Pregnancy  23.0    67.0   125.0    1.0
4  136      286.0  1st Pregnancy  25.0    62.0    93.0    0.0

In [13]:
df["smoke"].replace([0,1], ["NonSmoker", "Smoker"], inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  128      279.0  1st Pregnancy  28.0    64.0    77.0     Smoker
3  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
4  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker

## 6. Before we move forward with any statistics, let's identify and remove any outliers from the dataset. Using the IQR method, search for outliers in the 5 numeric variables. This will take some time and organization, be careful with your code! Make sure you keep track of how many rows/outliers are removed. 

In [14]:
df.shape

(1174, 7)

In [16]:
bwt_q1 = df["bwt"].quantile(.25)
bwt_q3 = df["bwt"].quantile(.75)
bwt_iqr = bwt_q3 - bwt_q1
bwt_top = bwt_q3 + (bwt_iqr * 1.5)
bwt_bot = bwt_q1 - (bwt_iqr * 1.5)
bwt_iqr_outliers = df.loc[(df["bwt"] > bwt_top) | (df["bwt"] < bwt_bot)].index

In [17]:
bwt_iqr_outliers

Int64Index([ 220,  287,  335,  430,  465,  492,  519,  557,  594,  663,  692,
             700,  780,  851,  859,  922,  964,  978, 1005, 1007, 1041, 1081,
            1090, 1111],
           dtype='int64')

In [18]:
df.loc[[220,  287,  335,  430,  465,  492,  519,  557,  594,  663,  692, 700,  780,  851,  859,  922,  964,  978, 1005, 1007, 1041, 1081, 1090, 1111]]

bwt  gestation                parity   age  height  weight      smoke
220   173      293.0         1st Pregnancy  30.0    63.0   110.0  NonSmoker
287    71      281.0         1st Pregnancy  32.0    60.0   117.0     Smoker
335    71      234.0         1st Pregnancy  32.0    64.0   110.0     Smoker
430    68      223.0         1st Pregnancy  32.0    66.0   149.0     Smoker
465    69      232.0         1st Pregnancy  31.0    59.0   103.0     Smoker
492    71      277.0         1st Pregnancy  40.0    69.0   135.0  NonSmoker
519   174      281.0         1st Pregnancy  37.0    67.0   155.0  NonSmoker
557   170      303.0  Multiple Pregnancies  21.0    64.0   129.0  NonSmoker
594   176      293.0  Multiple Pregnancies  19.0    68.0   180.0  NonSmoker
663   166      299.0         1st Pregnancy  26.0    68.0   140.0  NonSmoker
692   167      288.0  Multiple Pregnancies  19.0    63.0   117.0  NonSmoker
700   174      288.0         1st Pregnancy  25.0    61.0   182.0  NonSmoker
780    62      228.0         1st Pregnancy  24.0    61.0   107.0  NonSmoker
851    72      271.0         1st Pregnancy  39.0    61.0   136.0  NonSmoker
859    58      245.0         1st Pregnancy  34.0    64.0   156.0     Smoker
922    55      204.0         1st Pregnancy  35.0    65.0   140.0  NonSmoker
964   169      296.0         1st Pregnancy  33.0    67.0   185.0  NonSmoker
978    65      237.0         1st Pregnancy  31.0    67.0   130.0  NonSmoker
1005   73      277.0         1st Pregnancy  29.0    65.0   145.0  NonSmoker
1007   65      232.0         1st Pregnancy  24.0    66.0   125.0     Smoker
1041  174      284.0         1st Pregnancy  39.0    65.0   163.0  NonSmoker
1081   63      236.0  Multiple Pregnancies  24.0    58.0    99.0  NonSmoker
1090   72      266.0  Multiple Pregnancies  25.0    66.0   200.0     Smoker
1111   71      254.0         1st Pregnancy  19.0    61.0   145.0     Smoker

In [19]:
df = df.drop(bwt_iqr_outliers)

df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  128      279.0  1st Pregnancy  28.0    64.0    77.0     Smoker
3  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
4  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker

In [ ]:
##  Gestation

In [20]:
gest_q1 = df["gestation"].quantile(.25)
gest_q3 = df["gestation"].quantile(.75)
gest_iqr = gest_q3 - gest_q1
gest_top = gest_q3 + (gest_iqr * 1.5)
gest_bot = gest_q1 - (gest_iqr * 1.5)
gest_iqr_outliers = df.loc[(df["gestation"] > gest_top) | (df["gestation"] < gest_bot)].index

In [21]:
gest_iqr_outliers

Int64Index([   5,    6,    9,   56,   60,   63,  109,  119,  143,  173,  177,
             182,  193,  198,  200,  215,  220,  232,  237,  256,  319,  343,
             364,  406,  425,  445,  470,  482,  584,  632,  645,  654,  670,
             688,  701,  708,  717,  723,  767,  768,  771,  806,  858,  883,
             899,  930,  947,  952,  996, 1055, 1060, 1062, 1066, 1071, 1090,
            1095, 1114, 1121, 1122, 1131, 1133, 1140],
           dtype='int64')

In [22]:
df.loc[[5,    6,    9,   56,   60,   63,  109,  119,  143,  173,  177, 182,  193,  198,  200,  215,  220,  232,  237,  256,  319,  343, 364,  406,  425,  445,  470,  482,  584,  632,  645,  654,  670, 688,  701,  708,  717,  723,  767,  768,  771,  806,  858,  883, 899,  930,  947,  952,  996, 1055, 1060, 1062, 1066, 1071, 1090, 1095, 1114, 1121, 1122, 1131, 1133, 1140]]

bwt  gestation                parity   age  height  weight      smoke
5     138      244.0         1st Pregnancy  33.0    62.0   178.0  NonSmoker
6     132      245.0         1st Pregnancy  23.0    65.0   140.0  NonSmoker
9     140      351.0         1st Pregnancy  27.0    68.0   120.0  NonSmoker
56     75      232.0         1st Pregnancy  33.0    61.0   110.0  NonSmoker
60    101      245.0         1st Pregnancy  23.0    63.0   130.0     Smoker
...   ...        ...                   ...   ...     ...     ...        ...
1121  127      242.0         1st Pregnancy  17.0    61.0   135.0     Smoker
1122   87      247.0  Multiple Pregnancies  18.0    66.0   125.0     Smoker
1131  146      319.0         1st Pregnancy  28.0    66.0   145.0  NonSmoker
1133  110      321.0         1st Pregnancy  28.0    66.0   180.0  NonSmoker
1140  109      244.0  Multiple Pregnancies  21.0    63.0   102.0     Smoker

[62 rows x 7 columns]

In [23]:
df = df.drop(gest_iqr_outliers)

df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  128      279.0  1st Pregnancy  28.0    64.0    77.0     Smoker
3  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
4  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker

In [ ]:
## Age

In [24]:
age_q1 = df["age"].quantile(.25)
age_q3 = df["age"].quantile(.75)
age_iqr = age_q3 - age_q1
age_top = age_q3 + (age_iqr * 1.5)
age_bot = age_q1 - (age_iqr * 1.5)
age_iqr_outliers = df.loc[(df["age"] > age_top) | (df["age"] < age_bot)].index

In [25]:
age_iqr_outliers

Int64Index([557, 944], dtype='int64')

In [26]:
df.loc[[557, 944]]

bwt  gestation                parity   age  height  weight      smoke
557  143      294.0         1st Pregnancy  44.0    65.0   145.0  NonSmoker
944  122      280.0  Multiple Pregnancies  45.0    62.0   128.0  NonSmoker

In [27]:
df = df.drop(age_iqr_outliers)

df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  128      279.0  1st Pregnancy  28.0    64.0    77.0     Smoker
3  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
4  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker

In [ ]:
## Height

In [28]:
height_q1 = df["height"].quantile(.25)
height_q3 = df["height"].quantile(.75)
height_iqr = height_q3 - height_q1
height_top = height_q3 + (height_iqr * 1.5)
height_bot = height_q1 - (height_iqr * 1.5)
height_iqr_outliers = df.loc[(df["height"] > height_top) | (df["height"] < height_bot)].index

In [29]:
height_iqr_outliers

Int64Index([378, 1062], dtype='int64')

In [30]:
df.loc[[378, 1062]]

bwt  gestation         parity   age  height  weight   smoke
378   146      263.0  1st Pregnancy  39.0    53.0   110.0  Smoker
1062  141      281.0  1st Pregnancy  29.0    54.0   156.0  Smoker

In [31]:
df = df.drop(height_iqr_outliers)

df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  128      279.0  1st Pregnancy  28.0    64.0    77.0     Smoker
3  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
4  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker

In [ ]:
## Weight

In [32]:
weight_q1 = df["weight"].quantile(.25)
weight_q3 = df["weight"].quantile(.75)
weight_iqr = weight_q3 - weight_q1
weight_top = weight_q3 + (weight_iqr * 1.5)
weight_bot = weight_q1 - (weight_iqr * 1.5)
weight_iqr_outliers = df.loc[(df["weight"] > weight_top) | (df["weight"] < weight_bot)].index

In [33]:
weight_iqr_outliers

Int64Index([   2,   10,   17,   19,   23,   36,   78,  101,  108,  129,  131,
             140,  157,  158,  160,  189,  244,  266,  269,  298,  357,  372,
             398,  441,  443,  453,  457,  490,  521,  533,  547,  610,  611,
             633,  642,  743,  752,  753,  759,  778,  800,  811,  821,  888,
            1016, 1029, 1064, 1081],
           dtype='int64')

In [34]:
df.loc[[2,   10,   17,   19,   23,   36,   78,  101,  108,  129,  131, 140,  157,  158,  160,  189,  244,  266,  269,  298,  357,  372, 398,  441,  443,  453,  457,  490,  521,  533,  547,  610,  611, 633,  642,  743,  752,  753,  759,  778,  800,  811,  821,  888, 1016, 1029, 1064, 1081]]

bwt  gestation                parity   age  height  weight      smoke
2     128      279.0         1st Pregnancy  28.0    64.0    77.0     Smoker
10    114      273.0         1st Pregnancy  30.0    63.0   498.0  NonSmoker
17    115      274.0         1st Pregnancy  27.0    67.0   175.0     Smoker
19    122      276.0         1st Pregnancy  30.0    68.0   182.0  NonSmoker
23    114      266.0         1st Pregnancy  20.0    65.0   175.0     Smoker
36    110      278.0         1st Pregnancy  23.0    63.0   177.0  NonSmoker
78    125      305.0         1st Pregnancy  22.0    70.0   196.0     Smoker
101   131      283.0         1st Pregnancy  25.0    67.0   215.0  NonSmoker
108   115      283.0         1st Pregnancy  25.0    61.0  1500.0     Smoker
129   160      300.0         1st Pregnancy  29.0    71.0   175.0     Smoker
131   126      282.0         1st Pregnancy  38.0    66.0   250.0  NonSmoker
140   125      262.0         1st Pregnancy  36.0    66.0   190.0  NonSmoker
157   136      285.0         1st Pregnancy  23.0    62.0   175.0  NonSmoker
158   113      277.0         1st Pregnancy  23.0    65.0   192.0     Smoker
160   124      277.0         1st Pregnancy  29.0    63.0   220.0  NonSmoker
189   119      286.0         1st Pregnancy  20.0    64.0   180.0  NonSmoker
244   148      279.0         1st Pregnancy  27.0    71.0   189.0  NonSmoker
266   124      292.0         1st Pregnancy  29.0    68.0   176.0     Smoker
269   100      277.0         1st Pregnancy  31.0    62.0   355.0     Smoker
298   105      276.0         1st Pregnancy  22.0    67.0   845.0  NonSmoker
357   116      276.0         1st Pregnancy  33.0    61.0   180.0  NonSmoker
372   140      251.0         1st Pregnancy  28.0    63.0   210.0  NonSmoker
398    85      258.0         1st Pregnancy  41.0    67.0   399.0  NonSmoker
441   119      271.0         1st Pregnancy  28.0    64.0   175.0     Smoker
443   131      263.0         1st Pregnancy  29.0    64.0   180.0     Smoker
453   132      282.0         1st Pregnancy  28.0    67.0   200.0     Smoker
457   114      293.0  Multiple Pregnancies  20.0    66.0   180.0     Smoker
490   105      260.0         1st Pregnancy  23.0    64.0   197.0  NonSmoker
521   104      270.0         1st Pregnancy  25.0    61.0   511.0  NonSmoker
533   115      273.0  Multiple Pregnancies  23.0    67.0   215.0     Smoker
547    91      248.0         1st Pregnancy  33.0    63.0   202.0  NonSmoker
610   150      286.0         1st Pregnancy  38.0    67.0   175.0  NonSmoker
611    86      284.0         1st Pregnancy  39.0    65.0   174.0     Smoker
633   116      280.0         1st Pregnancy  34.0    68.0   198.0  NonSmoker
642    84      279.0         1st Pregnancy  34.0    63.0   190.0  NonSmoker
743   160      271.0         1st Pregnancy  32.0    67.0   215.0  NonSmoker
752   136      291.0         1st Pregnancy  41.0    66.0   191.0  NonSmoker
753   119      286.0         1st Pregnancy  22.0    63.0   185.0     Smoker
759   122      273.0         1st Pregnancy  26.0    66.0   210.0  NonSmoker
778   125      301.0  Multiple Pregnancies  35.0    68.0   181.0  NonSmoker
800   138      280.0  Multiple Pregnancies  30.0    65.0   478.0  NonSmoker
811   136      288.0         1st Pregnancy  23.0    62.0   217.0  NonSmoker
821    97      279.0         1st Pregnancy  29.0    68.0   178.0     Smoker
888   143      285.0         1st Pregnancy  27.0    68.0   185.0  NonSmoker
1016  117      284.0         1st Pregnancy  25.0    66.0   177.0     Smoker
1029  104      267.0         1st Pregnancy  30.0    63.0   180.0  NonSmoker
1064  138      286.0  Multiple Pregnancies  28.0    68.0    27.0  NonSmoker
1081  130      291.0         1st Pregnancy  30.0    65.0    67.0     Smoker

In [35]:
df = df.drop(weight_iqr_outliers)

df.reset_index(drop = True, inplace = True)

df.head()

bwt  gestation         parity   age  height  weight      smoke
0  120      284.0  1st Pregnancy  27.0    62.0   100.0  NonSmoker
1  113      282.0  1st Pregnancy  33.0    64.0   135.0  NonSmoker
2  108      282.0  1st Pregnancy  23.0    67.0   125.0     Smoker
3  136      286.0  1st Pregnancy  25.0    62.0    93.0  NonSmoker
4  120      289.0  1st Pregnancy  25.0    62.0   125.0  NonSmoker

In [36]:
df.shape

(1036, 7)

7. Describe the characteristics of your qualitative variables by doing the following:

    * Determine the frequencies of each group within each categorical variable. Which groups have the highest frequency?
    * Determine the relative frequencies of each group within each categorical variable. Which groups have the highest relative frequencies?
    * Create a crosstab table using both the categorical variables in your dataset. How many mothers are smokers who have been pregnant before?

In [37]:
df["parity"].value_counts()

## more mothers are in their 1st pregnancy

1st Pregnancy           754
Multiple Pregnancies    282
Name: parity, dtype: int64

In [38]:
df["parity"].value_counts(normalize = True)

## 

1st Pregnancy           0.727799
Multiple Pregnancies    0.272201
Name: parity, dtype: float64

In [39]:
df["smoke"].value_counts()

## there are more nonsmokers

NonSmoker    631
Smoker       405
Name: smoke, dtype: int64

In [40]:
df["smoke"].value_counts(normalize = True)

NonSmoker    0.609073
Smoker       0.390927
Name: smoke, dtype: float64

In [41]:
pd.crosstab(df["parity"], df["smoke"])

## 110 mothers are smokers who have had previous pregnancies

smoke                 NonSmoker  Smoker
parity                                 
1st Pregnancy               459     295
Multiple Pregnancies        172     110

8. Describe the Measures of Central Tendency (mean, median, mode) of the numeric variables in the dataset. Calculate the mean, median, and mode for all 5 numeric variables. For each median and mode, describe the following:

    * What does the median value tell you about the data?
    * What does the mode value tell you about the data?

In [42]:
print(df["bwt"].mean())
print(df["bwt"].median())
print(df["bwt"].mode())

## the values are close. there is a normal distribution in the data

120.17567567567568
120.0
0    117
dtype: int64


In [43]:
print(df["gestation"].mean())
print(df["gestation"].median())
print(df["gestation"].mode())

## the values are close. there is a normal distribution in the data

279.984555984556
280.0
0    282.0
dtype: float64


In [44]:
print(df["age"].mean())
print(df["age"].median())
print(df["age"].mode())

## the values are close. there is a normal distribution in the data

27.159266409266408
26.0
0    24.0
dtype: float64


In [45]:
print(df["height"].mean())
print(df["height"].median())
print(df["height"].mode())

## the values are close. there is a normal distribution in the data

64.00965250965251
64.0
0    65.0
dtype: float64


In [46]:
print(df["weight"].mean())
print(df["weight"].median())
print(df["weight"].mode())

## the values are close. there is a normal distribution in the data

125.65637065637065
125.0
0    130.0
dtype: float64


9. Describe the Measures of Variability of the numeric variables in the dataset. Calculate the range, standard deviation, and 85th percentile for all 5 numeric variables. Answer the following questions:

    * Which column has the largest range? Which has the smallest? What does this mean?
    * Which column has the largest standard deviation? Which has the smallest? What does this tell you about those variables?

In [50]:
bwt_range = df["bwt"].max() - df["bwt"].min()
bwt_std = df["bwt"].std()
bwt_85 = np.percentile(df["bwt"], 85)

print(bwt_range)
print(bwt_std)
print(bwt_85)

90
16.40437160501985
137.0


In [51]:
gest_range = df["gestation"].max() - df["gestation"].min()
gest_std = df["gestation"].std()
gest_85 = np.percentile(df["gestation"], 85)

print(gest_range)
print(gest_std)
print(gest_85)

64.0
11.396403911826766
292.0


In [52]:
age_range = df["age"].max() - df["age"].min()
age_std = df["age"].std()
age_85 = np.percentile(df["age"], 85)

print(age_range)
print(age_std)
print(age_85)

28.0
5.723814227977458
34.0


In [53]:
height_range = df["height"].max() - df["height"].min()
height_std = df["height"].std()
height_85 = np.percentile(df["height"], 85)

print(height_range)
print(height_std)
print(height_85)

16.0
2.481215121299967
67.0


In [54]:
weight_range = df["weight"].max() - df["weight"].min()
weight_std = df["weight"].std()
weight_85 = np.percentile(df["weight"], 85)

print(weight_range)
print(weight_std)
print(weight_85)

90.0
16.581045809853826
145.0


In [ ]:
## bwt and weight have the largest range. Most amount of variation between the values
## height has the smallest range. Least amount of variation between the values

In [ ]:
## weight has the largest SD. The data is more spread out around the average.  Height has the smallest SD. the data is more closely spread around the mean